### PR TITLE
1214 Migrate the quick tests to run as part of the integration tests

### PR DESCRIPTION
**Pull request by carlosfelix2**
I have been slowly working on this over a long period of time.  This PR converts the quick tests from a shell script using the PRIME CLI to a set of integration tests in Kotlin.  The ProcessData class, which is the one used for the PRIME CLI data process is reused programmatically and I added the reuse of the metadata object, so it runs quicker.  With this change the quick tests will also run as part of the build.

Test steps:
1. Run the integration tests using `./gradlew testIntegration` and verify the quick tests run successfully. 

## Changes
- Migrated the quick tests to the integration test suite.

### Testing
- [x] Tested locally?
- [ ] Ran `quickTest all`?  <------ THIS GOES AWAY!!!!!!
- [ ] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
- #1214

### CODE REVIEW

**Review by cweagans — COMMENTED** [review]
Since quicktest is no longer needed, should https://github.com/CDCgov/prime-reportstream/blob/045ffcef27fd180a4ba9d4e40b3e5f43295887e7/.github/pull_request_template.md?plain=1#L13 be removed from the PR template as part of this PR?

**Review by jimduff-usds — APPROVED** [review]
Very Nice.    I like that it still runs without needing to have ReportStream running.

I wonder that we couldn't call the underlying functions directly, instead of calling ProcessData.main, but I guess this way it acts as a basic test of some CLI capabilities, which is a good thing.

Is there a mode where I can see more about what it's doing while its running- eg, what states/schemas/etc?   I tried various verbose options, but didn't seem to find what I was looking for.

File: prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+ *   Output format - the output format (CSV, HL7, HL7_BATCH)
+ *   Name format - (OPTIONAL) a filename name format
+ *   Recv organization - (OPTIONAL) a receiving organization for the fake data
+ *   Output filename pattern - The pattern to test for when testing the output filename
+
+/**
+ * Any extra arguments to be passed in to the data processor separated by spaces.
+*/
+EXTRA_ARGS("Extra args
+ */
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt:30]: Fixed

File: prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    /**
+     * Read the configuration file.
+     * @return the list of configurations
+     */
+    private fun readTestConfig(): List<TestConfig> {
+        val config: List<TestConfig>
+        // Note we can only use input streams since the file may be in a JAR
+        val resourceStream = this::class.java.getResourceAsStream(configFile)
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt:111]: I am divided about this.  For the translation tests there are external files used, so I wanted to decouple the code from the configuration.  For this test we could just add it to the code, but I figure to decouple the code from the config as well.  No one should be touching the code after this I think.

File: prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            dataGenerator.main(args)
+        }
+
+        override fun execute() {
+            generateData()
+            assertThat(dataGenerator.outputReportFiles.size).isEqualTo(1)
+            QuickTestUtils.checkFilename(dataGenerator.outputReportFiles[0], config.expectedOutputFilePattern)
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt:194]: I have not had any issues with newline at end of files.  Will leave that as a separate styling issue.

**Comment by carlosfelix2** [comment]
> Very Nice. I like that it still runs without needing to have ReportStream running.
> 
> I wonder that we couldn't call the underlying functions directly, instead of calling ProcessData.main, but I guess this way it acts as a basic test of some CLI capabilities, which is a good thing.
> 
> Is there a mode where I can see more about what it's doing while its running- eg, what states/schemas/etc? I tried various verbose options, but didn't seem to find what I was looking for.

This code runs as a unit test, so there is no verbose option, although there is output from Process Data shown in the console as before.

File: prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosfelix2) [reply prime-router/src/testIntegration/kotlin/quicktests/QuickSchemaTranslationTests.kt:111]: Oh yes, I remember now.  It was to allow others like from the sender automation team to add more tests without having to change code.

**Comment by carlosfelix2** [comment]
> Very Nice. I like that it still runs without needing to have ReportStream running.
> 
> I wonder that we couldn't call the underlying functions directly, instead of calling ProcessData.main, but I guess this way it acts as a basic test of some CLI capabilities, which is a good thing.
> 
> Is there a mode where I can see more about what it's doing while its running- eg, what states/schemas/etc? I tried various verbose options, but didn't seem to find what I was looking for.

I think what you want is `gradle testIntegration -Pshowtests`.  The Github Action build has this turned on by default.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -2,21 +2,36 @@ This PR ...
 
 **If you are suggesting a fix for a currently exploitable issue, please disclose the issue to the prime-reportstream team directly outside of GitHub instead of filing a PR, so we may immediately patch the affected systems before a disclosure. See [SECURITY.md/Reporting a Vulnerability](https://github.com/CDCgov/prime-reportstream/blob/master/SECURITY.md#reporting-a-vulnerability) for more information.**
 
+Test Steps:
+1. *Include steps to test these changes*
+
```
Comment (ronaldheft-usds) [inline .github/pull_request_template.md:7]: The reorder of the template looks good. This appears like a more logical order.

**Review by ronaldheft-usds — APPROVED** [review]

**Pull request merged**